pkg/networkservice/common/heal/client.go

```diff
@@ -70,61 +72,79 @@ func NewClient(client networkservice.MonitorConnectionClient, onHeal *networkser
 		updateExecutor:    serialize.NewExecutor(),
 		eventReceiver:     nil, // This is intentionally nil
 		recvEventExecutor: serialize.NewExecutor(),
+		chainContext:      ctx,
 	}
+
 	if rv.onHeal == nil {
 		rv.onHeal = addressof.NetworkServiceClient(rv)
 	}
-	rv.updateExecutor.AsyncExec(func() {
-		runtime.SetFinalizer(rv, func(f *healClient) {
-			f.updateExecutor.AsyncExec(func() {
-				if f.cancelFunc != nil {
-					f.cancelFunc()
-				}
-			})
-		})
-		ctx, cancelFunc := context.WithCancel(context.Background())
-		rv.cancelFunc = cancelFunc
-		// TODO decide what to do about err here
-		recv, _ := rv.client.MonitorConnections(ctx, nil)
-		rv.eventReceiver = recv
-		rv.recvEventExecutor.AsyncExec(rv.recvEvent)
-	})
+
+	rv.init()
 
 	return rv
 }
 
+func (f *healClient) init() {
+	if f.eventReceiver != nil {
+		select {
+		case <-f.eventReceiver.Context().Done():
+			return
+		default:
+			f.eventReceiver = nil
+		}
+	}
+
+	logrus.Info("Creating new eventReceiver")
+
+	recv, err := f.client.MonitorConnections(f.chainContext, &networkservice.MonitorScopeSelector{}, grpc.WaitForReady(true))
+	if err != nil {
+		f.updateExecutor.AsyncExec(func() {
+			f.init()
+		})
+		return
+	}
+
+	f.eventReceiver = recv
+	f.recvEventExecutor.AsyncExec(f.recvEvent)
+}
+
 func (f *healClient) recvEvent() {
 	select {
 	case <-f.eventReceiver.Context().Done():
-		f.eventReceiver = nil
+		return
 	default:
 		event, err := f.eventReceiver.Recv()
-		if err != nil {
-			event = nil
-		}
 		f.updateExecutor.AsyncExec(func() {
+			if err != nil {
+				for id := range f.reported {
+					f.requestors[id]()
+					delete(f.reported, id)
+				}
+				f.init()
+				return
+			}
+
 			switch event.GetType() {
 			case networkservice.ConnectionEventType_INITIAL_STATE_TRANSFER:
 				f.reported = event.GetConnections()
+				if event.GetConnections() == nil {
+					f.reported = map[string]*networkservice.Connection{}
+				}
+
 			case networkservice.ConnectionEventType_UPDATE:
 				for _, conn := range event.GetConnections() {
 					f.reported[conn.GetId()] = conn
 				}
+
 			case networkservice.ConnectionEventType_DELETE:
-				for _, conn := range event.GetConnections() {
+				for id, conn := range event.GetConnections() {
+					f.requestors[id]()
 					delete(f.reported, conn.GetId())
-					if f.requestors[conn.GetId()] != nil {
-						f.requestors[conn.GetId()]()
-					}
-				}
-			}
-			for id, request := range f.requestors {
```

Comment:
We seem to have lost this logic somewhere in init().  What was the thought?

Reply:
Imagine the case:
1. Request(conn-2)
2. f.requestors[conn-2] = func() {...}
3. UPDATE from monitor for conn-1 (that has been created a time ago)
In that case `reported` and `requested` maps are inconsistent. And we will call `requestFunc()` for conn-2, but this is unnecessary

Reply:
So the case I think I'm not seing in the init code is: 

4.  What if we get an INIT_STATE_TRANSFER that is missing some of the requested?

Reply:
Requesting connections that are missing in INIT_STATE_TRANSFER might be excessive.  

If INIT_STATE_TRANSFER doesn't have any connections, then there are 2 options:
1. They will come in the following UPDATE event
2. Remote server is dead 

So probably in case of `err` from Recv() we should call all funcs in `requestors` map?

Reply:
If the remote server restarts, then we may well get and empty INIT_STATE_TRANSFER as our first hint it died... in which case we need to rerequest all the connections we have to for this client (and the sooner the better ;) )

Reply:
Yes, that's what I'm talking about. 
```go
event, err := f.eventReceiver.Recv()
if err != nil {
    for id, r := range f.requestors {
        r()
        delete(f.reported, id)
    }
    f.init()
    return
}
```
should be something like this, right?

Reply:
That would work :)

Reply:
@edwarnicke fixed and covered with tests